### PR TITLE
Remove support for Python 3.5

### PR DESCRIPTION
**Pull request by rbarry82**
Without support for < 3.6 needed, `__setname__` can be used freely,
and we are not required to have all `Object`s inherit from a
metaclass which can implement it.

This frees the Operator Framework to add metaclasses to other
types if needed without worrying about MRO when combining

### CODE REVIEW

**Review by jnsgruk — APPROVED** [review]
LGTM. Tested by building the latest Prometheus and Alertmanager charms and seems to check out fine.

File: README.md
Comment (benhoyt) [inline README.md:108]: This should say 18.04 (bionic) now. The default there is Python 3.6, by the looks: https://packages.ubuntu.com/bionic/python3

I haven't confirmed that PyYAML is included in Ubuntu's cloud image for 18.04, though.

File: ops/framework.py
Comment (benhoyt) [inline ops/framework.py:238]: Out of interest, why is "from None" needed here?

File: ops/model.py
Comment (benhoyt) [inline ops/model.py:1133]: Good catch, thanks!

**Review by benhoyt — APPROVED** [review]
It'll be good to leave 3.5 behind. Are we sure we can drop Xenial now? What was the discussion around that? I seem to recall last time it was raised (several months ago) there were still some charmers who pushed back on not supporting Xenial VMs, but I could be remembering wrong.

**Comment by wgrant** [comment]
> It'll be good to leave 3.5 behind. Are we sure we can drop Xenial now? What was the discussion around that? I seem to recall last time it was raised (several months ago) there were still some charmers who pushed back on not supporting Xenial VMs, but I could be remembering wrong.

From the perspective of a very heavy internal Canonical charmer and user, it would be quite unfortunate if we had to run old charm forks for the second half of xenial's support cycle.  IMO it's not justifiable to drop 3.5 while xenial's still extremely widely used, and will remain so for years to come.

**Comment by jnsgruk** [comment]
> > It'll be good to leave 3.5 behind. Are we sure we can drop Xenial now? What was the discussion around that? I seem to recall last time it was raised (several months ago) there were still some charmers who pushed back on not supporting Xenial VMs, but I could be remembering wrong.
> 
> From the perspective of a very heavy internal Canonical charmer and user, it would be quite unfortunate if we had to run old charm forks for the second half of xenial's support cycle. IMO it's not justifiable to drop 3.5 while xenial's still extremely widely used, and will remain so for years to come.

So I guess the answer here would just be to continue pinning to `ops<=1.2.0` in `requirements.txt` which will work in the majority of cases. Are there any charms in particular where this won't work, and the charm *can't* run on `bionic` or `focal`?

**Comment by wgrant** [comment]
> > > It'll be good to leave 3.5 behind. Are we sure we can drop Xenial now? What was the discussion around that? I seem to recall last time it was raised (several months ago) there were still some charmers who pushed back on not supporting Xenial VMs, but I could be remembering wrong.
> > 
> > 
> > From the perspective of a very heavy internal Canonical charmer and user, it would be quite unfortunate if we had to run old charm forks for the second half of xenial's support cycle. IMO it's not justifiable to drop 3.5 while xenial's still extremely widely used, and will remain so for years to come.
> 
> So I guess the answer here would just be to continue pinning to `ops<=1.2.0` in `requirements.txt` which will work in the majority of cases. Are there any charms in particular where this won't work, and the charm _can't_ run on `bionic` or `focal`?

I'm concerned about what will happen when popular subordinates (such as landscape-client or telegraf) get rewritten in OF, and we're stuck using the old versions on existing xenial deployments forever.

**Comment by jameinel** [comment]
I think the right way to provide Xenial support is to cut an official LTS release of OF, and be very clear that it is the branch/version of OF that supports Xenial. And then charms that *do* need to cross that compatibility have a clear target for what they should develop against.

The fact that PyPI and many other tools are strongly EOLing Python 3.5 support is going to make it rather hard to continue with Python 3.5 for the lifetime of Xenial ESM. But having a clear "here is a supported series of OF that supports Xenial" is a good bridge.
And that LTS can become "if there are features that you *need* for a Xenial supporting charm, then engage and we can look at backporting it". Obviously that gets harder the longer the spread is in time, but also the expected chance of needing it also diminishes.
I would like to see the *next* release of OF support Xenial (since there are a fair number of improvements in the current development branch), and then this lands prior to the next 1.X release of OF.

Juju itself is doing the cut off at 3.0. So 2.9 will be our LTS release of Juju that will continue to support Xenial, but we will be dropping support for Xenial for Juju 3.0. 2.9 is already necessary as an LTS because of our upgrade strategy (to upgrade from any 2.X you will need to upgrade to 2.9 first, and then you can upgrade to 3.0). So we do have a clear "it will be supported for a significant period of time".

**Review by jameinel — COMMENTED** [review]
I would hold off on landing this for a bit, but I support the initiative.

The other thing that we really lamented the loss of when we picked up 3.5 support was f-strings. So that may also be something to consider cleaning up the code base. (being able to do `f"{var} and another {var2}"` instead of `"{} and another {}".format(var, var2)`)
The former being a lot more human readable about intent, and not needing to count the number of {} to figure out which variable you were referencing.

**Comment by jnsgruk** [comment]
Thanks @jameinel for your input here.

> The other thing that we really lamented the loss of when we picked up 3.5 support was f-strings.

Agreed, @rbarry82 can you do a scrub of the codebase and move any existing cases of `"blah {}".format(some_var)` to use f-strings where it increases readability (noting that in some places I think `format` still has its merits)

> I would like to see the _next_ release of OF support Xenial (since there are a fair number of improvements in the current development branch), and then this lands prior to the next 1.X release of OF.

I think the approach of landing the current changes that are pending release in trunk to create a `1.2.1` release is a good idea. As soon as that has been released, we'll merge this branch and release `1.3`. We'll then consider the `1.2` branch an LTS release, and backport changes where there are security concerns or specific feature requests that make sense to port.

@wgx does that work for you?

File: ops/framework.py
Comment (rbarry82) [reply ops/framework.py:238]: Essentially to head off excessive exception trees from core Python. `__setname__` (like some builtins) has a very, uh, hardcoded exception handler because of the way it looks it up in class attributes. See [here](https://bugs.python.org/issue28214). Without this, even `exc.__cause__` requires multiple levels of nesting in the testsuite, but `from None` keeps the "user-facing" output the same as it was in the previous versions of OF.

File: README.md
Comment (rbarry82) [reply README.md:108]: Nice catch. I'll update the version and verify against the cloud image.

**Comment by rbarry82** [comment]
I'll go ahead with f-striing-ification. We'll still end up with `.format()` in some places for readability, but we should be able to cover 80% or more of the codebase with f-strings.

**Comment by wgrant** [comment]
> Thanks @jameinel for your input here.
> 
> > The other thing that we really lamented the loss of when we picked up 3.5 support was f-strings.
> 
> Agreed, @rbarry82 can you do a scrub of the codebase and move any existing cases of `"blah {}".format(some_var)` to use f-strings where it increases readability (noting that in some places I think `format` still has its merits)
> 
> > I would like to see the _next_ release of OF support Xenial (since there are a fair number of improvements in the current development branch), and then this lands prior to the next 1.X release of OF.
> 
> I think the approach of landing the current changes that are pending release in trunk to create a `1.2.1` release is a good idea. As soon as that has been released, we'll merge this branch and release `1.3`. We'll then consider the `1.2` branch an LTS release, and backport changes where there are security concerns or specific feature requests that make sense to port.
> 
> @wgrant does that work for you?

I really don't like it, and IME it's a good bit cheaper to eschew a handful of modern features rather than create a very early LTS, but if you've decided not to care then that's fine.

Other libraries dropping 3.5 support is less bad because they and their ecosystems are substantially more mature, so users are less likely to need newer versions for bugfixes or basic feature support. But sounds like you're set on it.

**Comment by rbarry82** [comment]
It won't hurt us to be prepared, even for a later release, and freeing up the Metapackage cruft may give us more flexibility

**Comment by Vultaire** [comment]
Seems CI is blocked on the Py3.5 checks - as expected due to this patch obviously, but that probably needs to be tweaked as well for things going forward if we want to merge this.

**Comment by petevg** [comment]
> Seems CI is blocked on the Py3.5 checks - as expected due to this patch obviously, but that probably needs to be tweaked as well for things going forward if we want to merge this.

I expect it would be, though luckily it's simple to fix. Just edit operator/.github/workflows/ci.yaml, and remove refs to 3.5!

**Comment by rbarry82** [comment]
Actually, [they are already gone](https://github.com/canonical/operator/pull/607/files#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133ddR12). A patch is coming up momentarily which removes the very last references to 3.5, and bumps it in `setup.py`, which may make the builder happier.

**Comment by benhoyt** [comment]
We (John M, Ian B, Harry P, and I) discussed this in our standup today, and thought it may be a bit hasty to do this right away, but that a compromise would make sense. Waiting till April 2024 when Xenial end-of-lifes is too long (Python 3.5 went end of life a year ago in Sept 2020), but doing it right now just to ease development slightly seems hasty and may cause pain for our users.

So the thinking was we should time this to coincide with the release of Juju 3.0 -- probably either late this year or early 2022. We'd publish the last release of Operator Framework that still supported Python 3.5 at that point, and then merge this change in.

**Comment by jnsgruk** [comment]
After some fairly extensive discussions following this PR, I'm going to close this PR *for now*.

At this particular time, it would seem there is just about enough demand for us to maintain 3.5 compatibility, given the relatively small improvements to be gained from dropping it. This is of course a trade-off, and one for which the variables will change over time. We will almost certainly revisit this at some point, but not *right now*.

Keeping this MR up to date and merged will be quite a lot of toil while we work out the strategy for supporting this going forward, which will likely be a cross-team effort with charmcraft, juju and operator framework :)

**Pull request closed without merging**
